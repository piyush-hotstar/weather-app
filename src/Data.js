import React from 'react';
import { observer } from "mobx-react-lite";
import { Table } from 'react-bootstrap';
import Alert from 'react-bootstrap/Alert';
import Accordion from 'react-bootstrap/Accordion';
import Beatloader from 'react-spinners/BeatLoader';

function Data ( {store} ) {

    const center = {
        display: "flex",
        justifyContent: "center"
    }

  if(store.loading)
  {
      return (
        <div className="container" style={{marginTop: "400px", display: "flex", justifyContent: "center"}}>
            <Beatloader />
        </div>
      )
   
  }
  else {

        const complete = store.whole

        const hourly = complete.hourly.data;
        const currently = complete.currently;

        return (
            <div className="container">
                {
                    store.message !=null ?
                    <Alert variant={"danger"}>{store.message}</Alert>
                    : null
                }
                <h5 style={center}>Coordinates: {store.coordinates.latitude}, {store.coordinates.longitude}</h5>
                <h5 style={center}>Timezone: {store.whole.timezone}</h5>
                <h1 style={center} className="text-7xl">{currently.temperature} &deg;F</h1>
                <h5 style={center} className="text-3xl">{currently.summary}</h5>
                <Accordion defaultActiveKey="0">
                <Accordion.Item eventKey="0">
                <Accordion.Header>Hourly statistics</Accordion.Header>
                <Accordion.Body>
                <ol>
                    <Table>
                        
                            <thead>
                        <tr>
                            <td>Time: </td>
                            <td>Summary: </td>
                            <td>Temperature: </td>
                        </tr>
                        </thead>
                        <tbody>
                    {
                        
                        hourly.map((hour,index) => (
                            <tr key={index}>
                            <td>{store.convert(hour.time)}</td>
                            <td>{hour.summary}</td>
                            <td>{hour.temperature}</td>
                            </tr>
                        ))
                        
                    }
                    </tbody>
                    </Table>
                </ol>
                </Accordion.Body>
                </Accordion.Item>
                </Accordion>
            </div>
            
        )
    }
  
}

export default observer(Data);
