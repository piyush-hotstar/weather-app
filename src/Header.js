import React from 'react';
import { observer } from "mobx-react-lite";
import Button from 'react-bootstrap/Button';
//import 'bootstrap/dist/css/bootstrap.min.css';

function Header ({store}) {

  return (
    <div className="container">
        <Button onClick={() => store.fixed()}> Fixed Location </Button>
        <Button variant="info" sty onClick={() => store.current()}> Current Location </Button>
    </div>
    )
  
}

export default observer(Header);
