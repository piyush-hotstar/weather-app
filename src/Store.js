import {
    action,
    computed,
    makeObservable,
    observable,
    autorun,
} from "mobx";
import axios from 'axios';
import { act } from "react-dom/cjs/react-dom-test-utils.production.min";

/*
this.coord = {latitude: position.coords.lat, longitude: â€¦}
*/
class Store {
    button = 0;
    coordinates = {
        latitude: "59.337239",
        longitude: "18.062381"
    };
    loading = true;
    whole = {};
    message = "";

    constructor() {
        makeObservable(this, {
            button: observable,
            coordinates: observable,
            message: observable,
            buttonStatus: computed,
            latitudeStatus: computed,
            longitudeStatus: computed,
            loading: observable,
            whole: observable,
            fixed: action,
            current: action,
            convert: action,
            getDataFromLocal: action
        });
        autorun(this.getJson);
    }

    fixed = () => {
        this.loading=true;
        this.button = 1;
        this.message = null;
        this.coordinates = {latitude: "59.337239", longitude: "18.062381"}
    }

    current = () => {
        this.loading=true;
        this.message = null;
        this.button=2;
        navigator.geolocation.getCurrentPosition( (position) => {
            this.coordinates = {latitude: position.coords.latitude, longitude: position.coords.longitude}
        },
        (error) => {
          if (error.code === error.PERMISSION_DENIED)
            {
              this.message = "You need to allow Permission :'("
              this.button=3; 
            }
            else {
                this.message = "It seems you are offline. Hence showing you cached data."
                this.latitude = "59.337239";
                this.longitude = "18.062381";
                //this.button = 2;
            }
        }
        );
        this.loading=true;
    }

    get buttonStatus() {
        return this.button;
    }

    get latitudeStatus() {
        return this.coordinates.latitude;
    }

    get longitudeStatus() {
        return this.coordinateslongitude;
    }

    convert = (time) => {
        var date = new Date(time);

        return date.toString("MMM dd");
    }

    getDataFromLocal = () => {
        console.log("no internet ii")
        this.message = "Something went wrong. Showing you cached data."
        this.whole = JSON.parse(localStorage.getItem('localStorage'))
        console.log(this.whole)
        this.loading=false
        
    }

    getJson = async () => {

        if(!navigator.onLine) {
            
            this.getDataFromLocal();
            this.message = "No internet connection. Showing you cached data."
            console.log(this.loading);
        }
        else {

            try {
                //axios.defaults.baseURL = 'https://api.darksky.net';
                let response = await axios.get(`https://cors-anywhere.herokuapp.com/https://api.darksky.net/forecast/2bb07c3bece89caf533ac9a5d23d8417/${this.coordinates.latitude},${this.coordinates.longitude}`)
                this.whole = response.data
                localStorage.setItem('localStorage', (JSON.stringify(response.data)));
                this.loading = false
            }
            catch (err) {
                this.getDataFromLocal();
                
                //return;
        }

        }
        //this.button = 2;
    }
}

export default Store;