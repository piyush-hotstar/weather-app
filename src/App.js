import React from 'react';
import './App.css';
import Data from './Data';
import Header from './Header';
import { observer } from "mobx-react-lite";
import 'bootstrap/dist/css/bootstrap.min.css';

function App ({store}) {
    if(store.buttonStatus===0) {
      return (
        <Header store={store}/>
      )
    }
    else if(store.buttonStatus===3){
      return (
        <p>
          {store.message}
        </p>
      )
    }
    else {
      return (
        <Data store={store}/>
      )
    }
  
} 

export default observer(App);
