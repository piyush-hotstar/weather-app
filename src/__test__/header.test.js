import React from 'react';
import renderer from 'react-test-renderer';
import Store from '../Store';
import {render, screen, cleanup} from '@testing-library/react';
import Header from '../Header';
import Data from '../Data';
import App from '../App'

const store = new Store();

afterEach(() => {
    cleanup();
})

test('should render header component', () => {
    render(<Header />);
    const headerElement = screen.getByTestId('header-1');
    expect(headerElement).toBeInTheDocument();
})

test("matches header snapshot", () => {
    const tree  = renderer.create(<Header />).toJSON();
    expect(tree).toMatchSnapshot();
})

test("matches data snapshot", () => {
    const tree  = renderer.create(<Data store={store}/>).toJSON();
    expect(tree).toMatchSnapshot();
})

test("matches app snapshot", () => {
    const tree  = renderer.create(<App store={store}/>).toJSON();
    expect(tree).toMatchSnapshot();
})