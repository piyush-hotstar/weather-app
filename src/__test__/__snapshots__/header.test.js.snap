// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches app snapshot 1`] = `
<div
  className="container"
  data-testid="header-1"
>
  <button
    className="btn btn-primary"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "display": "inline",
        "justifyContent": "center",
        "marginLeft": "600px",
        "marginTop": "400px",
      }
    }
    type="button"
  >
     Fixed Location 
  </button>
  <button
    className="btn btn-info"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "display": "inline",
        "marginLeft": "50px",
        "marginTop": "400px",
      }
    }
    type="button"
  >
     Current Location 
  </button>
</div>
`;

exports[`matches data snapshot 1`] = `
<div
  className="container"
  style={
    Object {
      "display": "flex",
      "justifyContent": "center",
      "marginTop": "400px",
    }
  }
>
  <span
    className="css-1xdhyk6"
  >
    <span
      className="css-dfmyvi"
    />
    <span
      className="css-b9bjvj"
    />
    <span
      className="css-dfmyvi"
    />
  </span>
</div>
`;

exports[`matches header snapshot 1`] = `
<div
  className="container"
  data-testid="header-1"
>
  <button
    className="btn btn-primary"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "display": "inline",
        "justifyContent": "center",
        "marginLeft": "600px",
        "marginTop": "400px",
      }
    }
    type="button"
  >
     Fixed Location 
  </button>
  <button
    className="btn btn-info"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "display": "inline",
        "marginLeft": "50px",
        "marginTop": "400px",
      }
    }
    type="button"
  >
     Current Location 
  </button>
</div>
`;
